mogp_gpu/src/gp_gpu.hpp

```diff
@@ -0,0 +1,764 @@
+    if (i < N) {
+        log_diag = log(A[i * (N+1)]);
+    }
+
+    int laneIdx = i % WARP_SIZE;
+    int warpIdx = i / WARP_SIZE;
+
+    __syncthreads();
+    log_diag = warp_reduce_sum(log_diag);
+
+    if (laneIdx == 0) work[warpIdx] = log_diag;
+
+    __syncthreads();
+
+    log_diag = work[laneIdx];
+
+    if (warpIdx == 0) log_diag = warp_reduce_sum(log_diag);
+
+    if (i == 0) *result = 2.0 * log_diag;
+}
+
+//
+// The unused arguments are needed for the CUB implementation
+void sum_log_diag(int N, double *A, double *result, double *, size_t)
+{
+    // The number of thread blocks *must* be 1
+    // N can be at most 1024 (WARP_SIZE**2)
+    int thread_blocks = 1;
+    int threads_per_block = N;
+    sum_log_diag_kernel<<<thread_blocks, threads_per_block>>>(N, A, result);
+}
+
+#else
+// ----------------------------------------
+
+// Implementation of sum_log_diag using cub::DeviceReduce
+
+struct LogSq : public thrust::unary_function<double, double>
+{
+    __host__ __device__ double operator()(double x) const { return 2.0 * log(x); }
+};
+
+void sum_log_diag(int N, double *A, double *result, double *work, size_t work_size)
+{
+    auto transform_it = thrust::make_transform_iterator(A, LogSq());
+    auto sr = make_strided_range(transform_it, transform_it + N*N, N+1);
+
+    cub::DeviceReduce::Sum(work, work_size, sr.begin(), result, N);
+}
+// ----------------------------------------
+#endif // USE_SHUFFLE_SUM_IMPL
+
+
+// ----------------------------------------
+// ------- Some useful typedefs for vectors and matrices
+
+typedef int obs_kind;
+typedef typename Eigen::Matrix<REAL, Eigen::Dynamic, Eigen::Dynamic, Eigen::RowMajor> mat;
+typedef typename Eigen::Ref<Eigen::Matrix<REAL, Eigen::Dynamic, Eigen::Dynamic, Eigen::RowMajor> > mat_ref;
+typedef typename Eigen::Matrix<REAL, Eigen::Dynamic, 1> vec;
+typedef typename Eigen::Ref<Eigen::Matrix<REAL, Eigen::Dynamic, 1> > vec_ref;
+// ----------------------------------------
+// enum to allow the python code to select the type of "nugget"
+enum nugget_type {NUG_ADAPTIVE, NUG_FIT, NUG_FIXED};
+
+
+
+//////////////////////////////////////////////////////////////////////////////////////
+////////////////////////// The Gaussian Process  class ///////////////////////////////
+//////////////////////////////////////////////////////////////////////////////////////
+
+// By convention, members ending "_d" are allocated on the device
+class DenseGP_GPU {
+    // batch size (for allocation)
+    unsigned int testing_size;
+
+    // Number of training points, dimension of input
+    unsigned int n, D;
+
+    // inputs
+    mat inputs;
+
+    // targets
+    vec targets;
+
+    // flag for whether we have fit hyperparameters
+    bool theta_fitted;
+
+    // handle for CUBLAS calls
+    cublasHandle_t cublasHandle;
+
+    // handle for cuSOLVER calls
+    cusolverDnHandle_t cusolverHandle;
+
+    // inverse covariance matrix on device
+    thrust::device_vector<REAL> invQ_d;
```

Comment:
I ran into a few situations where caching the inverse rather than the factorized covariance matrix led to instabilities. I presume this is done here because GPU matrix multiplication can be better parallelized than double back substitution with the factorized matrix. It would be worth running the GPU version on the test case that I wrote to test for this problem in `test_GaussianProcess_predict_variance` to see if you have this problem.

Reply:
Yes, that test case does fail on the GPU version.   I think our feeling is that this is version 1 of the GPU code, that works in most situations, and this could be improved in future iterations.